service/matching/taskReader.go

```diff
@@ -421,3 +376,120 @@ func (tr *taskReader) newDispatchContext(isolationGroup string) (context.Context
 	}
 	return tr.cancelCtx, func() {}
 }
+
+func (tr *taskReader) dispatchSingleTaskFromBufferWithRetries(isolationGroup string, taskInfo *persistence.TaskInfo) (breakDispatchLoop bool) {
+	// retry loop for dispatching a single task
+	for {
+		breakDispatchLoop, breakRetryLoop := tr.dispatchSingleTaskFromBuffer(isolationGroup, taskInfo)
+		if breakRetryLoop {
+			return breakDispatchLoop
+		}
+	}
+}
+
+func (tr *taskReader) dispatchSingleTaskFromBuffer(isolationGroup string, taskInfo *persistence.TaskInfo) (breakDispatchLoop bool, breakRetries bool) {
+	task := newInternalTask(taskInfo, tr.completeTask, types.TaskSourceDbBacklog, "", false, nil, isolationGroup)
+	dispatchCtx, cancel := tr.newDispatchContext(isolationGroup)
+	timerScope := tr.scope.StartTimer(metrics.AsyncMatchLatencyPerTaskList)
+	err := tr.dispatchTask(dispatchCtx, task)
+	timerScope.Stop()
+	cancel()
+	if err == nil {
+		return false, true
+	}
+	if err == context.Canceled {
+		tr.logger.Info("Tasklist manager context is cancelled, shutting down")
+		return true, true
+	}
+	if err == context.DeadlineExceeded {
```

Comment:
we can also early terminate here 
```
if err != deadline exceeded {
   // line 491-494 goes here
}

// handle deadline exceeded

Reply:
uh... not sure I understood sorry?

Reply:
[This block](https://github.com/uber/cadence/blob/b2667129d5cfe630ea6f07d908c82b3e8487314c/service/matching/taskReader.go#L404-L489) is unnecessary level of indentation which can be avoided by handling DeadlineExceeded first. Updated my example above to make the early termination obvious.

Reply:
Yeah, this is messy. Let me see if I can make it better.

Reply:
I realised we were handling unknown errors in a way which I think was wrong (they were just assumed to be rate-limit errors, but that seems... naieve. So ended up creating a default catchall for unknown errors as I feel this is more readable